pstoreds/peerstore.go

```diff
@@ -42,6 +43,59 @@ func NewPeerstore(ctx context.Context, store ds.TxnDatastore, opts Options) (pst
 		return nil, err
 	}
 
-	ps := pstore.NewPeerstore(pstoremem.NewKeyBook(), addrBook, pstoremem.NewPeerMetadata())
+	keyBook, err := NewKeyBook(ctx, store, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	peerMetadata, err := NewPeerMetadata(ctx, store, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	ps := pstore.NewPeerstore(keyBook, addrBook, peerMetadata)
 	return ps, nil
 }
+
+// uniquePeerIds extracts and returns unique peer IDs from database keys.
+func uniquePeerIds(ds ds.TxnDatastore, prefix ds.Key, extractor func(result query.Result) string) (peer.IDSlice, error) {
+	var (
+		q       = query.Query{Prefix: prefix.String(), KeysOnly: true}
+		results query.Results
+		err     error
+	)
+
+	txn, err := ds.NewTransaction(true)
+	if err != nil {
+		return peer.IDSlice{}, err
+	}
+	defer txn.Discard()
+
+	if results, err = txn.Query(q); err != nil {
+		log.Error(err)
+		return peer.IDSlice{}, err
+	}
+
+	defer results.Close()
+
+	idset := make(map[string]struct{})
+	for result := range results.Next() {
+		k := extractor(result)
+		idset[k] = struct{}{}
+		//key := ds.RawKey(result.Key)
+		//idset[key.Parent().Name()] = struct{}{}
+	}
+
+	if len(idset) == 0 {
+		return peer.IDSlice{}, nil
```

Comment:
same

Reply:
missed just one!

Reply:
This one is intentional. There are no results, and no error, hence we return an empty slice ;-)